frontend-react/docs/test-conventions.md

```diff
@@ -0,0 +1,492 @@
+
+Automated tests are an integral part of ongoing maintenance, and they help to fill in the cracks that manual testing inevitably leaves behind.  Of course, that's not to say that manual testing has no purpose; manual testing and automated testing are both essential at increasing the surface area of overall test coverage.
+
+To that end, we should subject our test code to the same rigor and scrutiny that we would our source code; tests should be clean, deterministic, and readily understandable for the benefit of anyone working in the codebase, regardless of how long they've been working in it.
+
+### Tests should read like a manual for source code
+
+Tests are often one of the first places to look when dealing with familiar code because they offer insight into how specific pieces function in isolation.  Because of this, we should strive to make our tests read like a manual for the related source code: "given this input, I should expect this output" or more practically, "given these conditions, I should expect this behavior."  Most concepts in programming should be able to be boiled down into inputs and outputs, and tests are no different in that regard.
+
+```tsx
+function sum(...numbers: number[]) {
+    return numbers.reduce((total, value) => total + value, 0);
+}
+
+// bad - what does it mean to function correctly?
+describe("sum", () => {
+    test("functions correctly", () => {
+        expect(sum()).toEqual(0);
+
+        expect(sum(1)).toEqual(1);
+        expect(sum(1, 2)).toEqual(3);
+        expect(sum(1, 2, 3)).toEqual(6);
+        expect(sum(1, 10, 100, 1000)).toEqual(1111);
+    });
+});
+
+// good - use descriptions to explain what's actually happening
+describe("sum", () => {
+    describe("with no arguments", () => {
+        test("returns 0", () => {
+            expect(sum()).toEqual(0);
+        });
+    });
+
+    describe("with arguments", () => {
+        test("returns the sum of all the arguments", () => {
+            expect(sum(1)).toEqual(1);
+            expect(sum(1, 2)).toEqual(3);
+            expect(sum(1, 2, 3)).toEqual(6);
+            expect(sum(1, 10, 100, 1000)).toEqual(1111);
+        });
+    });
+})
+```
+
+> Why?
+>
+> Having clearer and more sensible descriptions calls attention to what the purpose of the test actually is, which makes it easier to follow along with the underlying logic or to group similarly intentioned tests together.
+
+
+### Tests should check behavior, not implementation
+
+Especially in the world of front-end code, it's especially important to test our code from a user's perspective -- this means our tests should be prioritizing checking behavior ("what happens") rather than implementation details ("how it happens").
+
+```tsx
+function LoginForm() {
+    function handleSubmit() {
+        // assume a POST is sent to create a user session
+    }
+    
+    return (
+        <form onSubmit={handleSubmit}>
+            <button type="submit">
+        </form>
+    );
+}
+```
```

Comment:
Is the test-side of this example missing?

Reply:
Hahah whoops, pushed up a outdated version of this file.  Fixed!